### PR TITLE
util/time: use millisecond resolution for coarse Instant

### LINKED ISSUE #2158: coarse monotonic time jumps back
Please answer these questions before submitting your issue. Thanks!

1. What version of Rust are you using (`rustc --version`)?
    UTC Build Time:  2017-08-10 01:18:39
2. What operating system and processor architecture are you using?
    centos 6.8
3. What did you do?
4. What did you expect to see?
5. What did you see instead?
D0811 20:08:47.573029559   12094 env_linux.c:77]             Warning: insecure environment read function 'getenv' used
2017/08/11 20:13:47.927 panic_hook.rs:85: [ERROR] thread 'raftstore-1' panicked 'system time jumped back, 11295.363911996 -> 11295.363911995' at "src/util/time.rs:158"
stack backtrace:

```
   0:     0x7ff4e5e6969e - backtrace::backtrace::libunwind::trace
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/libunwind.rs:54
                         - backtrace::backtrace::trace<closure>
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/mod.rs:70
   1:     0x7ff4e5e69e53 - backtrace::capture::{{impl}}::new
                        at /home/jenkins/workspace/build_tikv_master/go/src/github.com/pingcap/tikv/target/release/build/backtrace-acbd9f8faa22306c/out/capture.rs:79
   2:     0x7ff4e575c1cc - tikv::util::panic_hook::set_exit_hook::{{closure}}
                        at src/util/panic_hook.rs:84
   3:     0x7ff4e5f0be2a - std::panicking::rust_panic_with_hook
                        at /checkout/src/libstd/panicking.rs:550
   4:     0x7ff4e5f0bcc4 - std::panicking::begin_panic<collections::string::String>
                        at /checkout/src/libstd/panicking.rs:511
   5:     0x7ff4e5f0bbf9 - std::panicking::begin_panic_fmt
                        at /checkout/src/libstd/panicking.rs:495
   6:     0x7ff4e576d698 - tikv::util::time::{{impl}}::elapsed
                        at src/util/time.rs:0
   7:     0x7ff4e576c388 - tikv::util::time::{{impl}}::elapsed
                        at src/util/time.rs:72
                         - tikv::util::time::{{impl}}::is_slow
                        at src/util/time.rs:76
   8:     0x7ff4e55f1b19 - tikv::raftstore::store::store::{{impl}}::timeout<tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>,tikv::pd::async::client::RpcClient>
                        at /home/jenkins/workspace/build_tikv_master/go/src/github.com/pingcap/tikv/src/util/macros.rs:128
   9:     0x7ff4e54d0770 - mio::event_loop::{{impl}}::timer_process<tikv::raftstore::store::store::Store<tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>, tikv::pd::async::client::RpcClient>>
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.5.1/src/event_loop.rs:368
                         - mio::event_loop::{{impl}}::run_once<tikv::raftstore::store::store::Store<tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>, tikv::pd::async::client::RpcClient>>
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.5.1/src/event_loop.rs:300
  10:     0x7ff4e5576132 - mio::event_loop::{{impl}}::run<tikv::raftstore::store::store::Store<tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>, tikv::pd::async::client::RpcClient>>
                        at /home/jenkins/.cargo/registry/src/github.com-1ecc6299db9ec823/mio-0.5.1/src/event_loop.rs:247
                         - tikv::raftstore::store::store::{{impl}}::run<tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>,tikv::pd::async::client::RpcClient>
                        at /home/jenkins/workspace/build_tikv_master/go/src/github.com/pingcap/tikv/src/raftstore/store/store.rs:455
  11:     0x7ff4e54ae51f - tikv::server::node::{{impl}}::start_store::{{closure}}<tikv::pd::async::client::RpcClient,tikv::server::transport::ServerTransport<tikv::server::transport::ServerRaftStoreRouter, tikv::server::resolve::PdStoreAddrResolver>>
                        at /home/jenkins/workspace/build_tikv_master/go/src/github.com/pingcap/tikv/src/server/node.rs:308
                         - std::sys_common::backtrace::__rust_begin_short_backtrace<closure,()>
                        at /checkout/src/libstd/sys_common/backtrace.rs:136
  12:     0x7ff4e54b72fd - std::thread::{{impl}}::spawn::{{closure}}::{{closure}}<closure,()>
                        at /checkout/src/libstd/thread/mod.rs:364
                         - std::panic::{{impl}}::call_once<(),closure>
                        at /checkout/src/libstd/panic.rs:296
                         - std::panicking::try::do_call<std::panic::AssertUnwindSafe<closure>,()>
                        at /checkout/src/libstd/panicking.rs:454
  13:     0x7ff4e5f12faa - panic_unwind::__rust_maybe_catch_panic
                        at /checkout/src/libpanic_unwind/lib.rs:98
  14:     0x7ff4e5542886 - std::panicking::try<(),std::panic::AssertUnwindSafe<closure>>
                        at /checkout/src/libstd/panicking.rs:433
                         - std::panic::catch_unwind<std::panic::AssertUnwindSafe<closure>,()>
                        at /checkout/src/libstd/panic.rs:361
                         - std::thread::{{impl}}::spawn::{{closure}}<closure,()>
                        at /checkout/src/libstd/thread/mod.rs:363
                         - alloc::boxed::{{impl}}::call_box<(),closure>
                        at /checkout/src/liballoc/boxed.rs:648
  15:     0x7ff4e5f0a9b5 - alloc::boxed::{{impl}}::call_once<(),()>
                        at /checkout/src/liballoc/boxed.rs:658
                         - std::sys_common::thread::start_thread
                        at /checkout/src/libstd/sys_common/thread.rs:21
                         - std::sys::imp::thread::{{impl}}::new::thread_start
                        at /checkout/src/libstd/sys/unix/thread.rs:84
  16:     0x7ff4e4acdaa0 - start_thread
  17:     0x7ff4e4604aac - __clone
  18:                0x0 - <unknown>
```

### PR DESCRIPTION
**Pull request by overvenus**
It's a quick fix for #2158. `SlowTimer` should not panic again by using millisecond resolution for coarse Instant. 

Fix #2158

### CODE REVIEW

File: src/util/time.rs [hunk 1]
```diff
@@ -247,44 +236,74 @@ impl Instant {
...
         }
     }
-}
 
-impl PartialEq for Instant {
-    fn eq(&self, other: &Instant) -> bool {
-        self.get_timespec().eq(&other.get_timespec())
+    fn elapsed_duration_coarse(later: Timespec, earlier: Timespec) -> Duration {
+        let sec = later.sec - earlier.sec;
+        let later_nsec = later.nsec - later.nsec % NANOSECONDS_PER_MILLISECOND;
```
Comment (siddontang) [inline src/util/time.rs]: why not using milliseconds to check directly?

Comment (overvenus) [inline src/util/time.rs]: `Duration::new()` takes nanoseconds.

Comment (siddontang) [inline src/util/time.rs]: what I mean here is convert the time to the millisecond and compare directly.

File: src/util/time.rs [hunk 2]
```diff
@@ -247,44 +237,74 @@ impl Instant {
...
     }
 }
 
-impl Ord for Instant {
-    fn cmp(&self, other: &Instant) -> Ordering {
-        self.get_timespec().cmp(&other.get_timespec())
+impl PartialEq for Instant {
+    fn eq(&self, other: &Instant) -> bool {
+        match (*self, *other) {
+            (Instant::Monotonic(this), Instant::Monotonic(other)) |
```
Comment (siddontang) [inline src/util/time.rs]: any test to cover it?

File: src/util/time.rs [hunk 3]
```diff
@@ -247,44 +237,74 @@ impl Instant {
...
+            _ => false,
+        }
     }
 }
 
 impl PartialOrd for Instant {
     fn partial_cmp(&self, other: &Instant) -> Option<Ordering> {
-        self.get_timespec().partial_cmp(&other.get_timespec())
+        match (*self, *other) {
+            (Instant::Monotonic(this), Instant::Monotonic(other)) |
```
Comment (siddontang) [inline src/util/time.rs]: ditto

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/util/time.rs (hunk 2 above)
Comment (overvenus) [inline src/util/time.rs]: It's already covered. 
https://github.com/pingcap/tikv/pull/2165/files/6dfd8f12f8be1be1e6450147716333cb7df9ff2b#diff-5a3b02da8619a0100baffd3477a68b21R411

File: src/util/time.rs [hunk 4]
```diff
@@ -247,44 +237,74 @@ impl Instant {
...
+                   later.sec as f64 + later.nsec as f64 / NANOSECONDS_PER_SECOND as f64);
         }
     }
-}
 
-impl PartialEq for Instant {
-    fn eq(&self, other: &Instant) -> bool {
-        self.get_timespec().eq(&other.get_timespec())
+    fn elapsed_duration_coarse(later: Timespec, earlier: Timespec) -> Duration {
+        let later_ms = later.sec * MILLISECOND_PER_SECOND +
```
Comment (siddontang) [inline src/util/time.rs]: please add a comment here

File: src/util/time.rs [hunk 5]
```diff
@@ -247,44 +237,74 @@ impl Instant {
...
 
-impl Ord for Instant {
-    fn cmp(&self, other: &Instant) -> Ordering {
-        self.get_timespec().cmp(&other.get_timespec())
+impl PartialEq for Instant {
+    fn eq(&self, other: &Instant) -> bool {
+        match (*self, *other) {
+            (Instant::Monotonic(this), Instant::Monotonic(other)) |
+            (Instant::MonotonicCoarse(this), Instant::MonotonicCoarse(other)) => this.eq(&other),
+            _ => false,
```
Comment (siddontang) [inline src/util/time.rs]: cover here?

File: src/util/time.rs [hunk 6]
```diff
@@ -247,44 +237,74 @@ impl Instant {
...
 impl PartialOrd for Instant {
     fn partial_cmp(&self, other: &Instant) -> Option<Ordering> {
-        self.get_timespec().partial_cmp(&other.get_timespec())
+        match (*self, *other) {
+            (Instant::Monotonic(this), Instant::Monotonic(other)) |
+            (Instant::MonotonicCoarse(this), Instant::MonotonicCoarse(other)) => {
+                this.partial_cmp(&other)
+            }
+            // The Order of different types of Instants is meaningless.
+            _ => None,
```
Comment (siddontang) [inline src/util/time.rs]: ditto

**Comment by siddontang** [comment]
I suggest running coarse monotonic elapsed check more times in test.

**Pull request merged**